### DISCUSSION TITLE
DDS-Router via SSH port forwarding/tunneling

### DISCUSSION THREAD

**Discussion by magladko** [eProsima/DDS-Router#405] [Q&A]
Hi all!

I am trying to run remotely ROS2 topics and have them visible among themselves.
I use vpn connection with a lot of firewall configuration that I cannot modify. Yet I have the ssh channel open and tunneling is possible there as I tested with netcat.

But I can't really make the routers to actually pass topic data between themselves, they just acknowledge their existence which I can see in wireshark. The dump for a single round of communication is here (the rest is almost identical):
```
No.     Time           Source                Destination           Protocol Length Info
     13 0.497267568    127.0.0.1             127.0.0.1             TCP      450    57404 → 11666 [PSH, ACK] Seq=1 Ack=1 Win=512 Len=382 TSval=2703425035 TSecr=2703422035

Frame 13: 450 bytes on wire (3600 bits), 450 bytes captured (3600 bits) on interface any, id 0
Linux cooked capture v1
Internet Protocol Version 4, Src: 127.0.0.1, Dst: 127.0.0.1
Transmission Control Protocol, Src Port: 57404, Dst Port: 11666, Seq: 1, Ack: 1, Len: 382
Data (382 bytes)

0000  52 54 43 50 7e 01 00 00 56 38 00 00 92 2d 52 54   RTCP~...V8...-RT
0010  50 53 02 02 01 0f 01 0f 7c 9b d4 e1 94 90 01 00   PS......|.......
0020  00 00 09 01 08 00 69 bb 2a 65 15 63 dd 94 15 05   ......i.*e.c....
0030  10 01 00 00 10 00 00 01 00 c7 00 01 00 c2 00 00   ................
0040  00 00 01 00 00 00 00 03 00 00 15 00 04 00 02 02   ................
0050  00 00 16 00 04 00 01 0f 00 00 50 00 10 00 01 0f   ..........P.....
0060  7c 9b d4 e1 94 90 01 00 00 00 00 00 01 c1 07 80   |...............
0070  04 00 04 00 00 00 32 00 18 00 04 00 00 00 d4 e1   ......2.........
0080  f4 1c 00 00 00 00 00 00 00 00 00 00 00 00 c0 a8   ................
0090  01 77 32 00 18 00 04 00 00 00 d4 e1 f4 1c 00 00   .w2.............
00a0  00 00 00 00 00 00 00 00 00 00 ac 11 af 4e 31 00   .............N1.
00b0  18 00 04 00 00 00 d4 e1 f5 1c 00 00 00 00 00 00   ................
00c0  00 00 00 00 00 00 c0 a8 01 77 31 00 18 00 04 00   .........w1.....
00d0  00 00 d4 e1 f5 1c 00 00 00 00 00 00 00 00 00 00   ................
00e0  00 00 ac 11 af 4e 02 00 08 00 14 00 00 00 00 00   .....N..........
00f0  00 00 58 00 04 00 3f 0c 00 00 62 00 14 00 0e 00   ..X...?...b.....
0100  00 00 52 6f 75 74 65 72 5f 43 6c 69 65 6e 74 00   ..Router_Client.
0110  00 00 59 00 28 00 01 00 00 00 11 00 00 00 50 41   ..Y.(.........PA
0120  52 54 49 43 49 50 41 4e 54 5f 54 59 50 45 00 00   RTICIPANT_TYPE..
0130  00 00 07 00 00 00 53 49 4d 50 4c 45 00 00 01 00   ......SIMPLE....
0140  00 00 80 01 38 00 04 00 00 00 92 2d 92 2d 00 00   ....8......-.-..
0150  00 00 00 00 00 00 00 00 00 00 7f 00 00 01 82 bb   ................
0160  2a 65 55 3b b1 15 0c 00 00 00 00 00 00 00 40 11   *eU;..........@.
0170  00 00 00 00 00 00 00 00 00 00 00 00 00 00         ..............

No.     Time           Source                Destination           Protocol Length Info
     14 0.497283939    127.0.0.1             127.0.0.1             TCP      68     11666 → 57404 [ACK] Seq=1 Ack=383 Win=506 Len=0 TSval=2703425035 TSecr=2703425035

Frame 14: 68 bytes on wire (544 bits), 68 bytes captured (544 bits) on interface any, id 0
Linux cooked capture v1
Internet Protocol Version 4, Src: 127.0.0.1, Dst: 127.0.0.1
Transmission Control Protocol, Src Port: 11666, Dst Port: 57404, Seq: 1, Ack: 383, Len: 0
```

Here I attach the config files:

My local machine:
---
```
version: v3.0

participants:

  - name: DDS_LAN_A
    kind: local

  - name: Local_Router
    kind: wan
    connection-addresses:
      - ip: 127.0.0.1
        port: 11666
        transport: tcp

```

Remote machine (acting as a server):
---
```
version: v3.0

participants:

  - name: DDS_LAN_B
    kind: local

  - name: Remote_Router
    kind: wan
    listening-addresses:
      - ip: 127.0.0.1
        port: 11666
        transport: tcp

```

The ssh tunnel command that I use on my local machine is: `ssh -N -L 11666:localhost:11666 <user>@<remotehost>`

Additional note: I managed to make the setup work on the localhost only, setting different domains in local participants, but I still have no idea on how to make this connection work through ssh.

**Comment by juanlofer-eprosima** [comment]
Hi @magladko ,

We have never tested routing data through an SSH tunnel, have you maybe tried with another application to verify it is not a setup issue?
Also, I did not fully understand; are you trying to make connections within a VPN, or from outside to the VPN?

Best regards

**Comment by magladko** [comment]
Hi @juanlofer-eprosima,
As for the tunnel, I verified it with netcat, so I know the two-way connection can be established.
As for the DDS or ROS2, I don't know what other applications can I look at.

The situation is as follows, both machines are connected to a single VPN, but only the ssh port is available to establish communication (others are blocked by quite a complicated configuration of firewalls which I cannot change). So my only option of getting data sent between hosts is through port 22.

Both hosts run up to date versions of ROS2 Humble and DDS-Router (not in containers) on Ubuntu 22.04.

Best

**Comment by magladko** [comment]
Well just in case anyone would need it, I tried using the sshuttle and it work. Not entirely sure where the difference lies, yet unfortunately the connection I got was many seconds apart, only getting worse over time.
